kivy/core/window/window_sdl2.py

```diff
@@ -759,6 +775,21 @@ def mainloop(self):
             else:
                 Logger.trace('WindowSDL: Unhandled event %s' % str(event))
 
+    def _dispatch_drop_event(self, action, args):
+        if action == 'dropfile':
+            self.dispatch('on_drop_file', args[0])
+        elif action == 'dropbegin':
```

Comment:
I'm assuming no mouse events happen before this, hence the original issue!? However, could we not use e.g. touch or some other non-mouse input to drop stuff? In which case we should not update the mouse!?

Reply:
SDL does not report `mousemotion` events and therefore `mouse_pos` does not get updated before `dropxxx` events. On Linux `windowenter` is received first and then `dropxxx` events, but on Windows `dropxxx` event will be first and then `windowenter`. 

I don't know if non-mouse devices can trigger `dropxxx` events from SDL. Also, how to handle the case when app is paused and receives `dropxxx` events?

Reply:
Looking at the code in sdl2, there's no direct connection between a drop event and the mouse. However, it would be better if it was a mouse event that caused it to keep the mouse position updated. So, what you're doing is good, the only change I would make is to check if the mouse is actually in the window, and only update the position then.

Reply:
We can check if mouse cursor is within window bounds, but I don't know how could we check if the drop is caused by the mouse.

Reply:
We don't have to check whether it's caused by a mouse. The goal is to make sure the mouse is up to date. So if it's not caused by the mouse but the mouse is in the window, then it's ok if we still update it (although it'd probably keep the same value). The thing to avoid is updating the mouse when it's outside the window.

Reply:
Ok. I'll add the changes.

Reply:
Problem with "if within the window" is that on Windows user can drop a file on title bar and on window edges, but those are not within the window bounds so old `mouse_pos` remains.

Reply:
On the edge/title bar counts as being outside the window, right? Then maybe we need to update the mouse pos to `-1` when it exits the screen, so when we drop it on the edges the value will be negative indicating it wasn't dropped inside.

Reply:
How is this supposed to behave with custom title bar ? As in that there is no title bar or border

Reply:
If there's no title bar, then it can be dropped on the edge or bar, meaning it will be the same as dropping in the center of the window.

Reply:
sdl might take care of it by itself will test it once the or Is ready for the final merge

Reply:
Negative values for `mouse_pos` would be a breaking change.

We can add `is_mouse_pos_within_window` method to `WindowBase` and note in `on_drop_begin` about edges/title bar case on Windows.

Reply:
@p0lygun With custom title bar `mouse_pos` will be updated with correct value because title bar is within the window.

Reply:
I dislike adding a new function just for this. In the ideal API, we should not have gotten the pos from `mouse_pos`, but the drop event would have included the dropped position in the event dispatch. How about resetting it to zero when it leaves the window?

Reply:
I would go with dispatching it with `on_drop_begin` and updating `mouse_pos` if it is within the window. Resetting to zero would indicate a mouse_pos changed (within the window) but it didn't.

Reply:
And mouse pos value dispatched with `on_drop_begin` would be relative to window position (left-top corner without title bar).

Reply:
Yeah, that sounds good!

Reply:
I've discovered that if:
- the cursor is on the title bar or the edges
- mouse wheel scroll up/down
- `windowenter` is received
- `mousewheel` is received

In this case `mouse_pos` would be out of the window bounds, so I've added a test to skip `mousewheel` and `windowenter` if cursor is not within the window. This could be the issue in the SDL.

Reply:
To add, click on the title bar or mouse motion over it does not trigger any events.